src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/DocumentSymbol/DocumentSymbolEndpoint.cs

```diff
@@ -0,0 +1,103 @@
+
+using DocumentSymbol = VisualStudio.LanguageServer.Protocol.DocumentSymbol;
+
+[LanguageServerEndpoint(Methods.TextDocumentDocumentSymbolName)]
+internal class DocumentSymbolEndpoint : IRazorRequestHandler<DocumentSymbolParams, SumType<DocumentSymbol[], SymbolInformation[]>>, IRegistrationExtension
+{
+    private readonly ClientNotifierServiceBase _languageServer;
+    private readonly IRazorDocumentMappingService _documentMappingService;
+
+    public DocumentSymbolEndpoint(ClientNotifierServiceBase languageServer, IRazorDocumentMappingService documentMappingService)
+    {
+        _languageServer = languageServer ?? throw new ArgumentNullException(nameof(languageServer));
+        _documentMappingService = documentMappingService ?? throw new ArgumentNullException(nameof(documentMappingService));
+    }
+
+    public bool MutatesSolutionState => false;
+
+    public RegistrationExtensionResult? GetRegistration(VSInternalClientCapabilities clientCapabilities)
+    {
+        var options = new DocumentSymbolOptions()
+        {
+            WorkDoneProgress = false
+        };
+
+        return new RegistrationExtensionResult("documentSymbolProvider", new SumType<bool, DocumentSymbolOptions>(options));
+    }
+
+    public TextDocumentIdentifier GetTextDocumentIdentifier(DocumentSymbolParams request)
+        => request.TextDocument;
+
+    public async Task<SumType<DocumentSymbol[], SymbolInformation[]>> HandleRequestAsync(DocumentSymbolParams request, RazorRequestContext context, CancellationToken cancellationToken)
+    {
+        if (request is null)
+        {
+            throw new ArgumentNullException(nameof(request));
+        }
+
+        var documentContext = context.GetRequiredDocumentContext();
+        var delegatedParams = new DelegatedDocumentSymbolParams(
+            new VersionedTextDocumentIdentifier()
+            {
+                Uri = documentContext.Uri,
+                Version = documentContext.Version
+            },
+            request.TextDocument.GetProjectContext()
+        );
+
+        var response = await _languageServer.SendRequestAsync<DelegatedDocumentSymbolParams, SumType<DocumentSymbol[], SymbolInformation[]>?>(
+            RazorLanguageServerCustomMessageTargets.RazorDocumentSymbolEndpoint,
+            delegatedParams,
+            cancellationToken).ConfigureAwait(false);
+
+        if (!response.HasValue)
+        {
+            return new();
+        }
+
+        var codeDocument = await documentContext.GetCodeDocumentAsync(cancellationToken).ConfigureAwait(false);
+        var csharpDocument = codeDocument.GetCSharpDocument();
+
+        if (response.Value.TryGetFirst(out var documentSymbols))
+        {
+            foreach (var documentSymbol in documentSymbols)
+            {
+                if (_documentMappingService.TryMapToGeneratedDocumentRange(csharpDocument, documentSymbol.Range, out var newRange))
+                {
+                    documentSymbol.Range = newRange;
```

Comment:
I would have thought if we coulnd't map it, we shouldn't be returning it as a result, otherwise the dropdown will presumably show things that only appear in the generated code. I would expect this construct a new list containing the subset of `documentSymbols` that were mapped.

Same comment applies to line 96